Library/Homebrew/test/tap_spec.rb

```diff
@@ -495,10 +495,52 @@ def setup_completion(link:)
     expect(homebrew_foo_tap.config[:foo]).to be_nil
   end
 
-  describe "#each" do
+  describe ".each" do
     it "returns an enumerator if no block is passed" do
       expect(described_class.each).to be_an_instance_of(Enumerator)
     end
+
+    context "when the core tap is not installed" do
+      around do |example|
+        FileUtils.rm_rf CoreTap.instance.path
+        example.run
+      ensure
+        (CoreTap.instance.path/"Formula").mkpath
+      end
+
+      it "includes the core tap with the api" do
+        ENV.delete("HOMEBREW_NO_INSTALL_FROM_API")
+        expect(described_class.to_a).to include(CoreTap.instance)
+      end
+
+      it "omits the core tap without the api" do
+        ENV["HOMEBREW_NO_INSTALL_FROM_API"] = "1"
+        expect(described_class.to_a).not_to include(CoreTap.instance)
+      end
```

Comment:
```suggestion
      it "includes the core tap with the api", :with_api do
        expect(described_class.to_a).to include(CoreTap.instance)
      end

      it "omits the core tap without the api", :without_api do
        expect(described_class.to_a).not_to include(CoreTap.instance)
      end
```

It might make sense to add before spec tags for `:with_api` and `:without_api` so that we can avoid this boilerplate and make tests simpler. I also know that specifying `ENV["HOMEBREW_NO_INSTALL_FROM_API"] = "1"` is technically unnecessary but it does improve readability.

I this makes sense to everyone it can be handled as a follow-up in another PR.

Reply:
@apainintheneck makes sense to me! Might be nicer still to have e.g. `with_api` be implicit and only `without_api` is needed.

Reply:
Some of the test suite would break in that case I believe. We already automatically set `HOMEBREW_NO_INSTALL_FROM_API` before running tests. I'd rather be explicit even if it adds a small amount redundancy.

I guess another option would be to do something like `api: true` and `api: false`.

Reply:
> We should flip this and unset `HOMEBREW_NO_INSTALL_FROM_API` 

`HOMEBREW_NO_INSTALL_FROM_API` is currently only set because a bunch of tests would break otherwise, so yes, the goal should be to unset it for tests eventually.

Reply:
> It would be ideal to not have to set `HOMEBREW_NO_INSTALL_FROM_API` before running tests.

Agreed.

> If we had some way of being reasonably sure that we weren't making unexpected network requests, I'd feel more comfortable unsetting `HOMEBREW_NO_INSTALL_FROM_API` by default.

I think at some point we need to just bite the bullet and switch over. It's mildly concerning to me that the vast majority of our users do not have `HOMEBREW_NO_INSTALL_FROM_API` set but it's what our test suite does by default.

Addressing this seems higher priority than ensuring we don't accidentally add network requests as a result (which can be done after).

Reply:
> Addressing this seems higher priority than ensuring we don't accidentally add network requests as a result (which can be done after).

I'm worried that this will end up making a bunch of tests flaky because of hard to predict network requests.

Reply:
Flaky tests can be fixed later. I'm more concerned about having tests right now that provide false confidence because they are running in a minority/legacy configuration by default.